### ISSUE TITLE
_app.js getInitialProps does not fire on every route change as per docs

### ISSUE THREAD

**Issue by gregg-cbs** [vercel/next.js#23388]
### What version of Next.js are you using?

10.0.5

### What version of Node.js are you using?

12.18.3

### What browser are you using?

Firefox

### What operating system are you using?

Windows

### How are you deploying your application?

Locally

### Describe the Bug

In my _app.js file I add App.getInitialProps .

In the `getIntialProps` i do some checks on the req headers but it seems `getInitialProps` only fires on certain occasions and not on every route change.

### Expected Behavior

getInitialProps to fire on every route change.

Or to have some way to inspect every route change on server side without having to create a custom server. This is important for redirects.

### To Reproduce

Create the _app.js file. Add a console log in the getInitialProps.

Have some serverside rendered pages and some static pages - transition between them using the Link component and notice logging only happens on certain occasions.

Screen shot of docs - is this true? or does the documentation need to be updated?
![image](https://user-images.githubusercontent.com/57114335/112470133-f750fc00-8d72-11eb-9650-0c245dc3152a.png)

**Comment by filipesmedeiros** [comment]
> However, if getInitialProps is used in a custom _app.js, and the page being navigated to implements getServerSideProps, then getInitialProps will run on the server.

It shouldn't run on statically generated pages, is this what's happening?

**Comment by ijjk** [comment]
Closing as it sounds like this is occurring for statically generated pages (getStaticProps) as mentioned in https://github.com/vercel/next.js/issues/23388#issuecomment-814151879 which is expected. If this is not the case please provide a more complete reproduction and we can investigate further

**Comment by gregg-cbs** [comment]
Yes thanks!

So initially i thought it would run on every page and break static generation (because of the docs).
Then i realized it runs if a page has getServerSideProps and probably getInitialProps and that static generation pages are still intact which is awesome and that if I wanted the _app.js getInititalProps to run I needed to add the getServerSideProps function to the page i wanted it to run on.

I think it would be nice if the documentation mentioned that.
When starting out with a new framework and things like this don't work you are never sure if its you or the frameworks intentions and if the docs don't mention it, leaves you unsure - took me two to three days before I got a brain fart and figured it out.
